test/org/javarosa/xpath/expr/Scenario.java

```diff
@@ -0,0 +1,490 @@
+        FormEntryModel formEntryModel = new FormEntryModel(formDef);
+        FormEntryController formEntryController = new FormEntryController(formEntryModel);
+        return new Scenario(formDef, formEntryController);
+    }
+
+    /**
+     * Sets the value of the element located at the given xPath in the main instance.
+     * <p>
+     * This method supports an enhanced version of xpath with the following perks and
+     * limitations:
+     * <ul>
+     * <li>Only supports absolute xpaths</li>
+     * <li>Supports adding the index (zero-indexed) of a repeat instance by suffixing it between
+     * brackets. Example that would select the fourth instance of the <code>/foo/bar</code>
+     * repeat: <code>/foo/bar[3]</code></li>
+     * </ul>
+     * <p>
+     * This method ensures that all the repeat instances required by the given xPath
+     * exist. For example: /data/people[1]/name will make sure the second repeat for
+     * /data/people exists.
+     * <p>
+     * This method ensures that any required downstream change after the given value
+     * is set is triggered.
+     */
+    // TODO Make more of these, one for each data type, and use the correct IAnswerData type
+    public void answer(String xPath, String value) {
+        // the xPath could include repeat group indexes, meaning that we expect
+        // some repeat group to exists. We check that and create missing repeats
+        // where they're needed.
+        createMissingRepeats(xPath);
+
+        // The xPath must match a model element. This ensures we can resolve it.
+        TreeElement element = Objects.requireNonNull(resolve(xPath));
+
+        // We wrap the given value and set it through the formDef, which triggers
+        // any downstream change in dependant elements (elements that have a
+        // calculate formula depending on this field, or itemsets rendering this
+        // field's value as a choice)
+        formDef.setValue(new StringData(value), element.getRef(), true);
+    }
+
+    /**
+     * Jumps to the first question with the given name.
+     * <p>
+     * This method is part of a set of methods that address form navigation in a dynamic way,
+     * imitating real user interaction with the form:
+     * <ul>
+     * <li>{@link #jumpToFirst(String)}
+     * <li>{@link #answer(String)}
+     * <li>{@link #next()}
+     * <li>{@link #next(String)}
+     * <li>{@link #atTheEndOfForm()}
+     * </ul>
+     */
+    public Scenario jumpToFirst(String name) {
+        jumpToFirstQuestionWithName(name);
+        return this;
+    }
+
+    /**
+     * Answers the current question.
+     * <p>
+     * This method is part of a set of methods that address form navigation in a dynamic way,
+     * imitating real user interaction with the form:
+     * <ul>
+     * <li>{@link #jumpToFirst(String)}
+     * <li>{@link #answer(String)}
+     * <li>{@link #next()}
+     * <li>{@link #next(String)}
+     * <li>{@link #atTheEndOfForm()}
+     * </ul>
+     */
```

Comment:
This documentation is really nice, but I’d prefer to avoid the duplication. Maybe just have those 9 lines in one place? An ordinary comment above the group of methods, perhaps?